### PR TITLE
Add TF implementation of `XGLMModel`

### LINKED ISSUE #16422: Add TF implementation of XGLM model
# 🚀 Feature request

Add TF implementation of XGLM model

## Motivation

Narrow the gap between PT and TF implementations.

## Your contribution

I'd like to work on this.

### PR DESCRIPTION
**Pull request by stancld**
# What does this PR do?

Fixes #16422

This PR adds TF implementation of `XGLMModel`.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@gante @patil-suraj

### CODE REVIEW

File: tests/xglm/test_modeling_tf_xglm.py
Comment (gante) [inline tests/xglm/test_modeling_tf_xglm.py:391]: You can remove this test -- the `max_time` feature in generate is not implemented on the TF side :)

**Review by gante — COMMENTED** [review]
Hey @stancld 👋  I've made a quick review, looks solid 🔥 

The `resize_token_embeddings` has the same problem as GPT-J, so I'd ignore it for now. As for the PT-TF equivalence, I'd like to suggest to wait a day or two to allow for this PR (https://github.com/huggingface/transformers/pull/16557) to be merged -- it may help you pinpoint the differences with the updated error message :)

In any case, if you'd like help with the PT-TF test, let me know. I will do a thorough review after it is passing 💪

**Comment by stancld** [comment]
Hello @gante, I would like to kindly ask for help here, as I'm not fully sure, where the problem might be :/ Currently, it looks like all the weights are loaded properly by a TF model, but it seems there still must be a glitch :| (I'm still learning TF so I might have made a silly mistake O:] ) Thanks a lot in advance!

**Comment by stancld** [comment]
> > Hello @gante, I would like to kindly ask for help here, as I'm not fully sure, where the problem might be :/ Currently, it looks like all the weights are loaded properly by a TF model, but it seems there still must be a glitch :| (I'm still learning TF so I might have made a silly mistake O:] ) Thanks a lot in advance!
> 
> No worries :) Are you trying to run the equivalence test as it is, or are you running another script?

I've tried to run the equivalence tests as they are.

**Comment by gante** [comment]
@stancld a little forewarning: I'm having issues in need of debugging in one of my own PRs as well and I'm off tomorrow, so I don't expect to have an answer before next Tuesday :)

File: tests/xglm/test_modeling_tf_xglm.py
Comment (gante) [inline tests/xglm/test_modeling_tf_xglm.py:23]:
```suggestion
```

File: tests/xglm/test_modeling_tf_xglm.py
Comment (gante) [inline tests/xglm/test_modeling_tf_xglm.py:263]:
```suggestion
class TFXGLMModelTest(TFModelTesterMixin, unittest.TestCase):
```

File: src/transformers/models/xglm/modeling_tf_xglm.py
Comment (gante) [inline src/transformers/models/xglm/modeling_tf_xglm.py:536]:
```suggestion
        hidden_states = tf.cast(inputs_embeds, dtype=tf.float32) + positions
```

**Review by gante — COMMENTED** [review]
(changes for locally passing tf-pt equivalence test)

**Comment by gante** [comment]
Hey @stancld 👋 Apologies for my delay :) 

I've added two sets of small changes, to hopefully help with test issues:
1. removed `TFCoreModelTesterMixin` from the tests -- it is a slow test suite reserved for key models like `bert` or `gpt2`;
2. added an explicit cast inside a layer -- on my end, I got an exception related to a variable being in `half-precision. On CI, I see that there is no error, only a failed numerical test.

I see that it didn't solve the issue here, even though it is passing on my local env. Digging deeper 🔍

**Comment by gante** [comment]
Screenshot 2022-04-24 at 15 14 15

(passing in my env)

**Comment by gante** [comment]
I've spun up a machine with a GPU, it also passes if executed on a GPU. @stancld, does the test fail in your local environment, after the changes? 
- If it fails and you have a GPU, check if passes on CPU (e.g. `CUDA_VISIBLE_DEVICES="" RUN_SLOW=1 RUN_PT_TF_CROSS_TESTS=1 py.test -vv tests/xglm/test_modeling_xglm.py::XGLMModelTest::test_pt_tf_model_equivalence`)
- If it passes, it might be a CI env problem 🤔

**Comment by stancld** [comment]
> * RUN_SLOW=1 RUN_PT_TF_CROSS_TESTS=1 py.test -vv tests/xglm/test_modeling_xglm.py::XGLMModelTest::test_pt_tf_model_equivalence

Hi @gante, thanks a lot for the update. I can confirm equivalence tests pass fine on CPU.

**Comment by stancld** [comment]
It seems equivalence tests are passing after rebasing to the 'main' branch. There are some remaining three failing tests, however, the outcome is a bit random as they don't always fail 🤔 

```
FAILED tests/xglm/test_modeling_tf_xglm.py::TFXGLMModelTest::test_xglm_model_att_mask_past - tensorflow.python.framework.errors_impl.InvalidArgumentError: Expected 'tf.Tensor(False, shape=(), dtype=bool)' to be true. Summarized data: b''
FAILED tests/xglm/test_modeling_tf_xglm.py::TFXGLMModelTest::test_xglm_model_past - tensorflow.python.framework.errors_impl.InvalidArgumentError: Expected 'tf.Tensor(False, shape=(), dtype=bool)' to be true. Summarized data: b''
FAILED tests/xglm/test_modeling_tf_xglm.py::TFXGLMModelTest::test_xglm_model_past_large_inputs - tensorflow.python.framework.errors_impl.InvalidArgumentError: Expected 'tf.Tensor(False, shape=(), dtype=bool)' to be true. Summarized dat...
```

**Comment by gante** [comment]
> I can confirm equivalence tests pass fine on CPU.
@stancld have you tried them on GPU? If so, what was the outcome, and what GPU model do you have? I'm trying to figure out why the tests fail in some conditions :D

**Comment by stancld** [comment]
> > I can confirm equivalence tests pass fine on CPU.
> 
> @stancld have you tried them on GPU? If so, what was the outcome, and what GPU model do you have? I'm trying to figure out why the tests fail in some conditions :D

Haven't tried yet. I'll run tests on GPU later today and let you know what it looks like. :]

**Comment by gante** [comment]
@stancld after your most recent push, the equivalent tests passed on CI 🤔 Well, one less problem to consider.

**Comment by stancld** [comment]
Hi @gante, I cleaned a test file a little bit to match the other test files. I'm gonna add some integration tests later tonight and then I hope the PR will be ready for the final review :]

**Comment by stancld** [comment]
Hi @gante, actually not sure if any integration tests should be added. Can I ask you, please, to have a look if there's anything missing now?
Also, I've uploaded the TF checkpoint to the HF hub and it should work just fine. You can find a checkpoint [here](https://huggingface.co/Stancld/xglm-564M).

**Review by gante — COMMENTED** [review]
Hey @stancld 👋  First of all, apologies for my delay in the review 🙈  I think the model is almost ready to go! I've added a few minor comments, including an action point for me (to add the TF weights in the facebook repo)

Regarding integration tests: a potentially missing test is whether the model can make proper use of left padding (for instance, TFGPT2 was not doing it until recently). You can see an example test [here](https://github.com/huggingface/transformers/blob/main/tests/models/gpt2/test_modeling_tf_gpt2.py#L539)

As always, thank you for the excellent contribution <3

**Comment by gante** [comment]
Ah -- you should be able to get rid of the CI errors by rebasing with `main` 👍

File: src/transformers/models/xglm/modeling_tf_xglm.py
Comment (stancld) [reply src/transformers/models/xglm/modeling_tf_xglm.py:546]: @gante That's definitely a good idea! Try to adjust the code accordingly! 🚀

**Comment by stancld** [comment]
Hello @gante, thanks a lot for your feedback and tips! :] I added an integration test on left padding and found out it actually doesn't work properly and generates some gibberish. Try to dig deep to solve the problem :]

**Comment by stancld** [comment]
> Hey @stancld 👋 First of all, apologies for my delay in the review 🙈 I think the model is almost ready to go! I've added a few minor comments, including an action point for me (to add the TF weights in the facebook repo)
> 
> Regarding integration tests: a potentially missing test is whether the model can make proper use of left padding (for instance, TFGPT2 was not doing it until recently). You can see an example test [here](https://github.com/huggingface/transformers/blob/main/tests/models/gpt2/test_modeling_tf_gpt2.py#L539)
> 
> As always, thank you for the excellent contribution <3

Hello @gante, is there by change any MR fixing left-padding for GPT2? O:]

**Comment by gante** [comment]
Weights PR open here: https://huggingface.co/facebook/xglm-564M/discussions/1

(one detail needs double-checking, other than that it should be good to go)

**Comment by stancld** [comment]
Hello @gante, sorry for a bit of delay on this PR. We've been just finishing #16792. I'll try to fix all the issues here afterwards :]

**Comment by gante** [comment]
Weights merged -- for the 564M model, the others I will merge after this PR gets merged

(No worries about delays, take your time 👍 )

**Comment by stancld** [comment]
Ola @gante, I'm pleased to announce that all failing slow tests have been fixed and work fine now (including left-padding generation).

However, there's a problem with a `tf_model.h5` in the `facebook/xglm-574M`, as the model using those weights produces gibberish output. I can confirm, nonetheless, that everything works perfectly with my TF weights available in [`Stancld/xglm-564M](https://huggingface.co/Stancld/xglm-564M). 

cc: @patil-suraj @Rocketknight1

**Comment by gante** [comment]
> However, there's a problem with a tf_model.h5 in the facebook/xglm-574M, as the model using those weights produces gibberish output. I can confirm, nonetheless, that everything works perfectly with my TF weights available in [`Stancld/xglm-564M](https://huggingface.co/Stancld/xglm-564M).

😬  I'll check what went wrong with my conversion. You converted using the `from_pretrained(from_pt=True)`, then stored the weights with `.save_weights()`, correct?

**Comment by stancld** [comment]
> > However, there's a problem with a tf_model.h5 in the facebook/xglm-574M, as the model using those weights produces gibberish output. I can confirm, nonetheless, that everything works perfectly with my TF weights available in [`Stancld/xglm-564M](https://huggingface.co/Stancld/xglm-564M).
> 
> 😬 I'll check what went wrong with my conversion. You converted using the `from_pretrained(from_pt=True)`, then stored the weights with `.save_weights()`, correct?

I used `.save_pretrained(...)` instead of `.save_weights().`

**Comment by gante** [comment]
@stancld the conversion I pushed was built with the first version of the `pt-to-tf` CLI, which was not converting the model head. I've opened a PR with the updated weights, which also contain the model head 🙌 

There is still one minor problem, but it's outside the scope of this PR -- the PT weights are stored in `float16`, and our TF tools do not have support for half-precision (yet!). 

weights PR 👉  https://huggingface.co/facebook/xglm-564M/discussions/2

**Comment by gante** [comment]
They were merged -- can you check whether the weights work? :)

**Comment by stancld** [comment]
> They were merged -- can you check whether the weights work? :)

Yes, can confirm all tests (including slow ones) work!

**Comment by gante** [comment]
Awesome, so all that's left is a full review.

@Rocketknight1 @patil-suraj

File: src/transformers/models/xglm/modeling_tf_xglm.py
Comment (Rocketknight1) [inline src/transformers/models/xglm/modeling_tf_xglm.py:582]: Eager-only debugging is a good way to do this, I think!

**Review by Rocketknight1 — APPROVED** [review]
This seems like a high-quality PR! Nothing jumped out to me as wrong or inefficient in the main model code, and the tests seem solid, so if they're passing I think it's very likely to be a good implementation.

File: tests/models/xglm/test_modeling_tf_xglm.py
Comment (gante) [inline tests/models/xglm/test_modeling_tf_xglm.py:284]: 👍

**Review by gante — APPROVED** [review]
LGTM!

One detail -- now that I've enabled XLA generation for all models, this model seems to be failing the tests. Since it is failing the fast test, the procedure would be to:
1. Add an attribute like [this one](https://github.com/huggingface/transformers/blob/main/src/transformers/models/led/modeling_tf_led.py#L2338) to `TFXGLMForCausalLM `
2. Add XGLM to [this issue](https://github.com/huggingface/transformers/issues/17935) (I already did :) )

**Comment by gante** [comment]
Needs to be rebased with `main` again, the problems in CI were already fixed there :)

Going to ask @patil-suraj for a review, then we can merge it (and convert the remaining models!)

File: src/transformers/models/xglm/modeling_tf_xglm.py
Comment (patrickvonplaten) [inline src/transformers/models/xglm/modeling_tf_xglm.py:590]: Does layerdrop work in TF? Think this is a copy-paste / leftover from PyTorch... not sure if anyone has ever used layerdrop in TF. Fine by me to leave, but might be better to delete it here maybe?

File: src/transformers/models/xglm/modeling_tf_xglm.py
Comment (patrickvonplaten) [inline src/transformers/models/xglm/modeling_tf_xglm.py:877]:
```suggestion
        # TODO (Joao): investigate why XGLM has numerical issues in XLA generate
```

Let's directly open an issue for this maybe once the PR is merged. XGLM is an important model IMO

**Review by patrickvonplaten — APPROVED** [review]
Very cool @stancld !

**Comment by stancld** [comment]
@gante @patrickvonplaten One failing test fixed :]

**Comment by gante** [comment]
Splendid! Thank you for all the work, @stancld 🧡 

I'm going to merge the PR and push the weights for the other XGLM architectures 🔥

**Pull request merged**
